File: prob003/prob003.rb
primes = [2, 3, 5, 7, 11] #we already know some, so why not

def next_prime primes
  infinity = 1.0/0
  start = primes[-1]+1

  (start..infinity).each do |candidate|
    if primes.any? {|prime| candidate % prime == 0}
      # is not prime, skip
    else
      return candidate
    end
  end
end

num     = 600851475143
factors = []
i       = 0

while(num > 1) do
  if(i == primes.length)
    primes << next_prime(primes)
  end
  if(num % primes[i] == 0)
    factors << primes[i]
    num = num / primes[i]
  else
    i += 1
  end
end
puts factors.sort.last

File: prob003/prob003.clj
(def num-to-check 600851475143)

(defn next-prime [primes candidate]
  (if (some (fn [x] (= (rem candidate x) 0)) primes)
    (recur primes (inc candidate))
    candidate))




(defn factorize-4 [number primes divisor factors]
  (cond
    (= number 1) factors
    (= (rem number divisor) 0)
      (recur (quot number divisor) primes divisor (conj factors divisor))
    :else 
      (let [new-prime (next-prime primes (inc divisor))]
      (recur number (conj primes new-prime) new-prime factors))))

(defn factorize-2 [number primes]
  (factorize-4 number primes (first primes) []))

(def factorize
  (factorize-2 num-to-check [2]))

(def answer
  (first (into '() factorize)))


